### PR TITLE
Feature: event onClientWaterCannonHit

### PR DESCRIPTION
**Pull request by CrosRoad95**
Event: onClientWaterCannonHit
Arguments: x, y, z, normalx, normaly, normalz, materialid

Pr including hack, that allow to hit water cannon to hit water cannon custom objects created by createObject function.

image

Test resource: [testv2.zip](https://github.com/multitheftauto/mtasa-blue/files/2408147/testv2.zip)

Test https://www.youtube.com/watch?v=jPCnYIqQJoY&feature=youtu.be

and 4 reasons for marge
Fire cannon https://www.youtube.com/watch?v=um7A9IThtDU&feature=youtu.be
Grenade cannon https://www.youtube.com/watch?v=f2M-180uMZk&feature=youtu.be
Teleport cannon https://www.youtube.com/watch?v=0AmnWL5RsYY&feature=youtu.be
Paint cannon https://www.youtube.com/watch?v=u5eaHl_SGZQ&feature=youtu.be

Thanks for Saml1er and Botder for helping with hook.

### CODE REVIEW

File: Client/mods/deathmatch/logic/CClientGame.cpp [hunk 1]
```diff
@@ -6866,3 +6874,24 @@ void CClientGame::RemoveAnimationAssociationFromMap(CAnimBlendAssociationSAInter
...
+{
+    CLuaArguments Arguments;
+    Arguments.PushNumber(pColPoint.Position.fX);
+    Arguments.PushNumber(pColPoint.Position.fY);
+    Arguments.PushNumber(pColPoint.Position.fZ);
+    Arguments.PushNumber(pColPoint.Normal.fX);
+    Arguments.PushNumber(pColPoint.Normal.fY);
+    Arguments.PushNumber(pColPoint.Normal.fZ);
+    Arguments.PushNumber(pColPoint.ucSurfaceTypeB);
+    /*if (&pEntity != nullptr) // seems to not work
```
Comment (ghost) [inline Client/mods/deathmatch/logic/CClientGame.cpp]: Remove the comment block.

File: Client/multiplayer_sa/CMultiplayerSA.cpp [hunk 2]
```diff
@@ -1,6872 +1,6917 @@
...
-    SHeatHazeSettings settings;
-    settings.ucIntensity = 0x50;
-    settings.ucRandomShift = 0x0;
-    settings.usSpeedMin = 0x0C;
-    settings.usSpeedMax = 0x12;
-    settings.sScanSizeX = 0x4B;
-    settings.sScanSizeY = 0x50;
-    settings.usRenderSizeX = 0x50;
-    settings.usRenderSizeY = 0x55;
-    se
```
Comment (saml1er) [inline Client/multiplayer_sa/CMultiplayerSA.cpp]: why do you have two asm blocks? you can simply put this code after popad in first asm block.

File: Client/mods/deathmatch/logic/CClientGame.cpp [hunk 3]
```diff
@@ -6866,3 +6874,24 @@ void CClientGame::RemoveAnimationAssociationFromMap(CAnimBlendAssociationSAInter
 {
     m_mapOfCustomAnimationAssociations.erase(pAnimAssociation);
 }
+
+void CClientGame::WaterCannonHitWorldHandler(CColPointSAInterface& pColPoint, CEntitySAInterface& pEntity)
```
Comment (ghost) [inline Client/mods/deathmatch/logic/CClientGame.cpp]: what's the use of  `CEntitySAInterface& pEntity`? you don't seem to access it.

File: Client/multiplayer_sa/CMultiplayerSA.cpp (hunk 2 above)
Comment (ghost) [inline Client/multiplayer_sa/CMultiplayerSA.cpp]: why do you have two asm blocks? you can simply put this code after popad in first asm block.

**Comment by CrosRoad95** [comment]
fixed along Saml1er tips.
Removed conflicts
new test resource
and video with visualization how this working https://www.youtube.com/watch?v=jPCnYIqQJoY&feature=youtu.be

I think is ready to marge

**Comment by gatno** [comment]
Great work!

File: Client/multiplayer_sa/CMultiplayerSA.cpp [hunk 4]
```diff
@@ -300,6 +300,9 @@ DWORD dwFUNC_CAEVehicleAudioEntity__ProcessAIProp = FUNC_CAEVehicleAudioEntity__
 #define HOOKPOS_CTaskSimpleSwim_ProcessSwimmingResistance   0x68A4EF
 DWORD RETURN_CTaskSimpleSwim_ProcessSwimmingResistance = 0x68A50E;
 
+#define HOOKPOS_WatterCannonHitWorld   0x72932A
```
Comment (qaisjp) [inline Client/multiplayer_sa/CMultiplayerSA.cpp]: `Watter` should be `Water`

**Comment by botder** [comment]
Last commit renames event to `onClientElementHitByWaterCannon` with the following parameters:
- **vehicle** The vehicle with the mounted water cannon
- **hitX, hitY, hitZ** World-space hit position
- **normalX, normalY, normalZ** Surface normal
- **model** Model of the hit surface
- **materialID** Hit surface id

The **source** element is either the hit element or the root element for static world objects (you use **model** to identify it)

**Pull request merged**